File: hwacha_perf/trace.py
"""Spike（hlog 构建 + 本仓库补丁）指令级踪迹的解析，与 cc/src/trace.cc 同一格式。"""
from __future__ import annotations
import re
from dataclasses import dataclass, field
from typing import Optional


@dataclass
class TraceInstr:
    pc: int
    inst: int
    next: int
    vl: int
    scalar: bool
    active: list = field(default_factory=list)          # 每元素 0/1
    mem: list = field(default_factory=list)             # [(ut, addr)]

    def taken(self) -> bool:
        return self.next != self.pc + 8

    def active_at(self, e: int) -> bool:
        return self.scalar or e >= len(self.active) or bool(self.active[e])


@dataclass
class TraceBlock:
    start_pc: int
    vl: int
    instrs: list = field(default_factory=list)
    base_pc: Optional[int] = None     # 内核文件第一条指令对应的地址（多入口块时 != start_pc）

    def index_of(self, pc: int) -> int:
        return (pc - (self.base_pc if self.base_pc is not None else self.start_pc)) // 8


_WT = re.compile(r'^H: WT pc=([0-9a-f]+) inst=([0-9a-f]+) next=([0-9a-f]+) vl=(\d+) act=(\S*)')


def load_trace(path: str, lo: int = 0, hi: int = 1 << 64, blocks: Optional[tuple] = None,
               base: Optional[int] = None) -> list[TraceBlock]:
    """base：内核文件第一条指令的地址；不给则每个块以自身起点为基址（单入口块）。"""
    out: list[TraceBlock] = []
    cur: Optional[TraceBlock] = None
    expect = None
    pending: list = []

    def flush():
        nonlocal cur
        if cur is not None and cur.instrs and lo <= cur.start_pc < hi:
            out.append(cur)
        cur = None

    with open(path, errors='replace') as f:
        for line in f:
            if line.startswith('H: WT '):
                m = _WT.match(line)
                if not m:
                    continue
                pc, inst, nxt, vl = int(m.group(1), 16), int(m.group(2), 16), int(m.group(3), 16), int(m.group(4))
                act = m.group(5)
                if cur is None or pc != expect:
                    flush()
                    cur = TraceBlock(pc, vl, base_pc=base)
                ti = TraceInstr(pc, inst, nxt, vl, act == '-')
                if act != '-':
                    ti.active = [(int(act[i // 4], 16) >> (i % 4)) & 1 if i // 4 < len(act) else 0 for i in range(vl)]
                ti.mem = pending
                pending = []
                cur.instrs.append(ti)
                expect = nxt
                if (inst & 0xfff) == 0xc3f and (inst >> 12) == 0:
                    flush()
            elif line.startswith('HMEM:'):
                if line.startswith('HMEM: read from '):
                    addr = int(line[16:].split()[0], 16)
                elif line.startswith('HMEM: write '):
                    addr = int(line[12:].split()[0], 16)
                elif line.startswith('HMEM: rmw '):
                    addr = int(line[10:].split()[0], 16)
                else:
                    continue
                mu = re.search(r'ut=(\d+)', line)
                pending.append((int(mu.group(1)) if mu else 0, addr))
    flush()
    if blocks:
        out = out[blocks[0]:blocks[1]]
    return out

File: cc/src/trace.cc
#include "trace.hh"
#include "sim.hh"
#include <cstdio>
#include <cstring>
#include <fstream>

namespace hw {

static uint64_t hexField(const std::string &line, const char *key) {
    size_t p = line.find(key);
    if (p == std::string::npos) return 0;
    return std::strtoull(line.c_str() + p + std::strlen(key), nullptr, 16);
}
static unsigned decField(const std::string &line, const char *key) {
    size_t p = line.find(key);
    if (p == std::string::npos) return 0;
    return (unsigned)std::strtoul(line.c_str() + p + std::strlen(key), nullptr, 10);
}

Trace Trace::load(const std::string &path, uint64_t lo, uint64_t hi, uint64_t base) {
    std::ifstream f(path);
    if (!f) sim::fatal("cannot open trace " + path);
    Trace t;
    TraceBlock cur;
    bool inBlock = false;
    uint64_t expectPc = 0;
    std::vector<std::pair<unsigned, uint64_t>> pendingMem;
    std::string line;
    auto flush = [&]() {
        if (inBlock && !cur.instrs.empty() && cur.startPc >= lo && cur.startPc < hi) t.blocks.push_back(cur);
        cur = TraceBlock(); inBlock = false;
    };
    while (std::getline(f, line)) {
        if (line.rfind("H: WT ", 0) == 0) {
            TraceInstr ti;
            ti.pc = hexField(line, "pc=");
            ti.inst = hexField(line, "inst=");
            ti.next = hexField(line, "next=");
            ti.vl = decField(line, "vl=");
            size_t ap = line.find("act=");
            std::string act = ap == std::string::npos ? "-" : line.substr(ap + 4);
            while (!act.empty() && (act.back() == '\r' || act.back() == ' ')) act.pop_back();
            // 块边界：上一条是 vstop（next == 0 或未接续）或 pc 不等于上一条的 next
            if (!inBlock || ti.pc != expectPc) { flush(); inBlock = true; cur.startPc = ti.pc; cur.basePc = base ? base : ti.pc; cur.vl = ti.vl; }
            if (act == "-") ti.scalar = true;
            else {
                ti.active.assign(ti.vl, 0);
                for (unsigned i = 0; i < ti.vl; ++i) {
                    char c = i / 4 < act.size() ? act[i / 4] : '0';
                    unsigned nib = c <= '9' ? c - '0' : (c | 0x20) - 'a' + 10;
                    ti.active[i] = (nib >> (i % 4)) & 1;
                }
            }
            ti.mem.swap(pendingMem);
            cur.instrs.push_back(ti);
            expectPc = ti.next;
            // vstop：inst 低 12 位 0xc3f 且高位全零
            if ((ti.inst & 0xfff) == 0xc3f && (ti.inst >> 12) == 0) flush();
        } else if (line.rfind("HMEM:", 0) == 0) {
            uint64_t addr = 0;
            if (line.rfind("HMEM: read from ", 0) == 0) addr = std::strtoull(line.c_str() + 16, nullptr, 16);
            else if (line.rfind("HMEM: write ", 0) == 0) addr = std::strtoull(line.c_str() + 12, nullptr, 16);
            else if (line.rfind("HMEM: rmw ", 0) == 0) addr = std::strtoull(line.c_str() + 10, nullptr, 16);
            else continue;
            unsigned ut = decField(line, "ut=");
            pendingMem.push_back({ut, addr});
        }
    }
    flush();
    return t;
}

void Trace::filter(uint64_t lo, uint64_t hi) {
    std::vector<TraceBlock> keep;
    for (auto &b : blocks) if (b.startPc >= lo && b.startPc < hi) keep.push_back(b);
    blocks.swap(keep);
}

}  // namespace hw
